ResearchKit/Common/ORKTaskViewController.m

```diff
@@ -893,9 +893,18 @@ - (void)showViewController:(ORKStepViewController *)viewController goForward:(BO
         }
     }
     
-    if (step.identifier && ![_managedStepIdentifiers.lastObject isEqualToString:step.identifier]) {
-        [_managedStepIdentifiers addObject:step.identifier];
+    NSString *stepIdentifier = step.identifier;
+    if (stepIdentifier) {
+        NSUInteger indexOfStep = [_managedStepIdentifiers indexOfObject:stepIdentifier];
+        if (indexOfStep == NSNotFound) {
+            [_managedStepIdentifiers addObject:stepIdentifier];
+        } else {
+            // If the step being shown has been visited before, remove all the managed step identifiers visited after it
+            NSUInteger numberManagedOfStepIdentifiers = _managedStepIdentifiers.count;
+            [_managedStepIdentifiers removeObjectsInRange:NSMakeRange(indexOfStep + 1, numberManagedOfStepIdentifiers - (indexOfStep + 1))];
```

Comment:
A dictionary type `managedResults` stores the actual results, which is used to fill in user's previous answer on previously visited steps. A previous answer cannot be determined if we store multiple copies of answer for same step. Right now, I cannot think of a better way to replace the `managedResults`. Then returning duplicate results for all the times the user looped over is inevitable.

Being said, currently solution is not perfect. If there is a loop,  `managedResults` preserves user's path, but returning same results are for the same steps.

One solution could be to detect the loop within ORKTaskViewController and log warning for the developer and explain the consequences(returning duplicate results).